### PR TITLE
docs: add boot sequence proposal

### PR DESCRIPTION
**Pull request by andrewrynhard**
This is a proposal that introduces the notion of a boot sequencer.

### CODE REVIEW

File: docs/proposals/20190801-BootSequence.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+type Status int
+
+const (
+    Pass Status = iota
+    Fail
+)
+
+type Level struct {
+    status Status
+    phases []phase.Phase
```
Comment (smira) [inline docs/proposals/20190801-BootSequence.md]: should it be `phases [][]phase.Phase`?

File: docs/proposals/20190801-BootSequence.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    boot.Register(&Phase0Task{}, 0)
+}
+```
+
+Notice that `phase.Runner` is embedded.
+This serves to enforce the idea that a `BootSequencer` is a wrapper around a `phase.Runner`, that adds some extra functionality on top of it.
+The `BootSequencer` will perform the following tasks:
+
+- register phases and build the execution order
+- create any event handling channels
```
Comment (smira) [inline docs/proposals/20190801-BootSequence.md]: can we instead have a singleton-like standalone object act as service bus components can subscribe to and publish to?

e.g. ACPI listener can publish shutdown even there, while machined's main.go can listen on it to trigger shutdown?

does it have part of boot sequence, or we can think of that as standalone component?

Comment (andrewrynhard) [reply docs/proposals/20190801-BootSequence.md]: Yeah I like this idea. This object I think more or less will encompass all of what `machined` does. The main function could just set this thing up and then wait for a signal. Maybe this standalone component contains the boot sequencer.

File: docs/proposals/20190801-BootSequence.md (hunk 1 above)
Comment (andrewrynhard) [reply docs/proposals/20190801-BootSequence.md]: I think so ... building out an example go file and will be sure it compiles. I will copy over here when done.

**Comment by smira** [comment]
Another potential approach to the boot sequence and cleanup might be borrowing concept of typical Go HTTP middleware:

```go
type Phase func(somedata *Data, nextPhase phase) error

func examplePhase func(somedata *Data, nextPhase error) {
     if err := doSomething(); err != nil {
         return err
     }
     defer cleanup()

     return nextPhase(somedata)
}
```

in this case phases are totally free to implement cleanup the way they want, and cleanup happens "naturally" as `defer` objects are run during stack unwinding either via normal return sequence or due to the panic.

**Comment by andrewrynhard** [comment]
This is outdated, and not really worth salvaging.

**Pull request closed without merging**